### PR TITLE
Fix broken Qasm3 gates.

### PR DESCRIPTION
**Pull request by keriksson-rosenqvist**
Allow reuse of variable as long as it's the same variable. `old_var == var`

### CODE REVIEW

File: tests/qat/test_qasm.py [hunk 1]
```diff
@@ -603,6 +604,18 @@ def test_sech(self):
         assert np.isclose(pulses[1].amp, 0.5)
         assert np.isclose(pulses[1].std_dev, 20e-9)
 
+    @pytest.mark.parametrize(
+        "qasm", get_default_qasm3_gate_qasms(), ids=lambda val: val.split("\n")[-2]
+    )
+    def test_default_gates(self, qasm):
```
Comment (lcauser-oqc) [inline tests/qat/test_qasm.py:610]: Would it be possible to have a test with an QASM3 script that uses multiple of the problematic gate(s)?

**Review by lcauser-oqc — APPROVED** [review]
Thanks!

**Review by bgsach — APPROVED** [review]

File: src/qat/purr/integrations/qasm.py [hunk 2]
```diff
@@ -856,9 +856,10 @@ def _create_qb_specific_gate_suffix(
         return f"{name}[{','.join([str(qb) for qb in target_qubits])}]"
 
     def _attempt_declaration(self, var: Variable, context: QasmContext):
-        if var.name in context.variables:
+        if (old_var := context.variables.get(var.name, None)) is None:
+            context.variables[var.name] = var
+        elif old_var != var:
             raise ValueError(f"Can't redeclare variable {var.name}")
-        context.variables[var.name] = var
```
[the remark this answers is not in the document]
Comment (lcauser-oqc) [reply src/qat/purr/integrations/qasm.py:861]: TIL that `:=` has a name, and better yet, one that will make me laugh every time I use it :smile:

[the remark this answers is not in the document]
Comment (keriksson-rosenqvist) [reply src/qat/purr/integrations/qasm.py:861]: I'm OK either way. Walrus operators are generally something I like to for things that are very short lived/contained. I.e. if the `old_var` persisted past the if/elif context I'd probably declare it separately but as it's only used there I think it's fine.

File: tests/qat/qasm_utils.py [hunk 3]
```diff
@@ -64,3 +69,24 @@ def parse_and_apply_optimiziations(
 
     builder = parser.parse(get_builder(hardware), qasm)
     return builder
+
+
+qasm3_gates = {}
```
[the remark this answers is not in the document]
Comment (keriksson-rosenqvist) [reply tests/qat/qasm_utils.py:74]: For reuse, i.e. not having to parse the whole file again. Happy to make it `_qasm3_gates`

File: tests/qat/test_qasm.py [hunk 4]
```diff
@@ -603,6 +604,33 @@ def test_sech(self):
         assert np.isclose(pulses[1].amp, 0.5)
         assert np.isclose(pulses[1].std_dev, 20e-9)
 
+    @pytest.mark.parametrize(
+        "gate_tup", get_default_qasm3_gate_qasms(), ids=lambda val: val[-1]
+    )
+    def test_default_gates(self, gate_tup):
+        """Check that each default gate can be parsed individually."""
+        N, gate_string = gate_tup
+        qasm = f"OPENQASM 3.0;\nbit[{N}] c;\nqubit[{N}] q;\n{gate_string}\nmeasure q -> c;"
```
[the remark this answers is not in the document]
Comment (keriksson-rosenqvist) [reply tests/qat/test_qasm.py:613]: Not sure what you mean.
This is a parameterised test that where `gate_string` is a qasm string version of the each gate defined in `stdgates.inc` individually. The goal being that we know that any gate defined inside `stdgates.inc` can be parsed by our system.

**Pull request merged**